src/serac/numerics/assembled_sparse_matrix.cpp

```diff
@@ -0,0 +1,263 @@
+  // offsets are the row offsets corresponding to a vdof
+  // indices maps a given vdof to the the assembled element matrix vector (dof * ne + d).
+  // gatherMap takes an element matrix vector offset (dof, ne) and returns the partition-local vdof (d.o.f. id).
+  auto& test_offsets    = detail::ElementRestriction_offsets(test_restriction_);
+  auto& test_indices    = detail::ElementRestriction_indices(test_restriction_);
+  auto& trial_gatherMap = detail::ElementRestriction_gatherMap(trial_restriction_);
+
+  /**
+     We expect mat_ea to be of size (test_elem_dof * test_vdim, trial_elem_dof * trial_vdim, ne)
+     We assume a consistent striding from (elem_dof, vd) within each element
+  */
+  const int test_elem_dof  = test_fes_.GetFE(0)->GetDof();
+  const int trial_elem_dof = trial_fes_.GetFE(0)->GetDof();
+  const int test_vdim      = test_fes_.GetVDim();
+  const int trial_vdim     = trial_fes_.GetVDim();
+  const int test_ndofs     = test_fes_.GetNDofs();
+
+  for (int i = 0; i < test_vdim * test_ndofs; i++) {
+    I[i] = 0;
+  }
+
+  for (int test_vdof = 0; test_vdof < test_fes_.GetNDofs(); test_vdof++) {
+    // Look through each element corresponding to a test_vdof
+    const int test_row_offset = test_offsets[test_vdof];
+    const int nrow_elems      = test_offsets[test_vdof + 1] - test_row_offset;
+
+    // Build temporary array to get rid of duplicates
+    mfem::Array<int> trial_vdofs(nrow_elems * trial_elem_dof);
+    trial_vdofs = -1;
+    int nnz_row = 0;
+    for (int e_index = 0; e_index < nrow_elems; e_index++) {
+      // test_indices can be negative in the case of Hcurl
+      const int test_index_v = test_indices[test_row_offset + e_index];
+      const int test_index   = test_index_v >= 0 ? test_index_v : -test_index_v - 1;
+      const int e            = test_index / test_elem_dof;
+
+      // find corresponding trial_vdofs
+      mfem::Array<int> trial_elem_vdofs(trial_elem_dof);
+      for (int j = 0; j < trial_elem_dof; j++) {
+        // this might be negative
+        const auto trial_j_vdof_v = trial_gatherMap[trial_elem_dof * e + j];
+        const auto trial_j_vdof   = trial_j_vdof_v >= 0 ? trial_j_vdof_v : -1 - trial_j_vdof_v;
+        trial_elem_vdofs[j]       = trial_j_vdof;
+        if (trial_vdofs.Find(trial_j_vdof) == -1) {
+          // we haven't seen this before
+          trial_vdofs[nnz_row] = trial_j_vdof;
+          nnz_row++;
+        }
+      }
+    }
+
+    // add entries to I
+    for (int vi = 0; vi < test_vdim; vi++) {
+      const auto nnz_index_v = test_fes_.DofToVDof(test_vdof, vi);
+      const auto nnz_index   = nnz_index_v >= 0 ? nnz_index_v : -1 - nnz_index_v;
+      I[nnz_index]           = nnz_row * trial_vdim;
+    }
+  }
+
+  // Perform inclusive scan on all entries
+  int nnz = 0;
+  for (int i = 0; i < test_ndofs * trial_vdim; i++) {
+    int temp = I[i];
+    I[i]     = nnz;
+    nnz += temp;
+  }
+  I[test_ndofs * trial_vdim] = nnz;
```

Comment:
If `I[0]` is always zero, I think it's an exclusive scan. I have not tried it yet, but apparently C++17 introduced
https://en.cppreference.com/w/cpp/algorithm/exclusive_scan as well. If that fits our needs, I think it would be good to use the standard library algorithms when possible.

Reply:
It is always zero for hypre CSR I think. I didn't know there was an exclusive_scan available. Thanks

Reply:
it looks like these scan operations support execution space template arguments, so they may already support parallel execution. 

inclusive and exclusive scans were added as part of the Parallelism_TS to C++17 by Jared Hoberock, NVIDIA's main developer of thrust. So it sounds like he likely just added many of the parallel algorithms from thrust directly into the C++ standard!

Reply:
Interesting! Anyways I added a note in the code. since it seems that gcc 8.3.1 and clang 10.0.0 on LC don't support the scan operations unfortunately.